test/extensions/filters/network/kafka/codec_test.cc

```diff
@@ -0,0 +1,229 @@
+#include "extensions/filters/network/kafka/codec.h"
+
+#include "test/mocks/server/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+
+class RequestDecoderTest : public testing::Test {
+public:
+  Buffer::OwnedImpl buffer_;
+
+  template <typename T>
+  std::shared_ptr<T> serializeAndDeserialize(T request);
+};
+
+class MockMessageListener : public MessageListener {
+public:
+  MOCK_METHOD1(onMessage, void(MessageSharedPtr));
+};
+
+template <typename T>
+std::shared_ptr<T> RequestDecoderTest::serializeAndDeserialize(T request) {
+  char data[1024];
+  RequestSerializer serializer;
+  size_t written = serializer.encode(request, data);
+  buffer_.add(data, written);
+
+  std::shared_ptr<MockMessageListener> mock_listener = std::make_shared<MockMessageListener>();
+  RequestDecoder testee{RequestParserResolver::KAFKA_0_11, { mock_listener }};
+
+  MessageSharedPtr receivedMessage;
+  EXPECT_CALL(*mock_listener, onMessage(_))
+    .WillOnce(testing::SaveArg<0>(&receivedMessage));
+
+  testee.onData(buffer_);
+
+  return std::dynamic_pointer_cast<T>(receivedMessage);
+};
+
+// === FETCH (1) ============================================================
+
+TEST_F(RequestDecoderTest, shouldParseFetchRequest) {
+  // given
+  FetchRequest request{
+    1,
+    1000,
+    10,
+    20,
+    2,
+    {{
+        {
+            "topic1",
+            {{ { 10, 20, 1000, 2000 } }}
+        },
+        {
+            "topic1",
+            {{ { 11, 21, 1001, 2001 }, { 12, 22, 1002, 2002 } }}
+        },
+        {
+            "topic1",
+            {{ { 13, 23, 1003, 2003 } }}
+        },
+    }}
+  };
+  request.apiVersion() = 5;
+  request.correlationId() = 10;
+  request.clientId() = "client-id";
+
+  // when
+  auto received = serializeAndDeserialize(request);
+
+  // then
+  ASSERT_NE(received, nullptr);
+  ASSERT_EQ(*received, request);
+}
+
+// === LIST OFFSETS (2) ========================================================
+
+TEST_F(RequestDecoderTest, shouldParseListOffsetsRequest) {
+  // given
+  ListOffsetsRequest request{
+    10,
+    2,
+    {{
+        { "topic1", {{ {1, 1000}, {2, 2000} } }},
+        { "topic2", {{ {3, 3000} } }},
+    }}
+  };
+  request.apiVersion() = 2;
+  request.correlationId() = 10;
+  request.clientId() = "client-id";
+
+  // when
+  auto received = serializeAndDeserialize(request);
+
+  // then
+  ASSERT_NE(received, nullptr);
+  ASSERT_EQ(*received, request);
+}
+
+// === METADATA (3) ============================================================
+
+TEST_F(RequestDecoderTest, shouldParseMetadataRequest) {
+  // given
+  MetadataRequest request{
+    {{ "t1", "t2", "t3" }},
+    true
+  };
+  request.apiVersion() = 4;
+  request.correlationId() = 10;
+  request.clientId() = "client-id";
```

Comment:
the API for RequestConstruction is not perfect
especially considering that some versions are structurally identical, so it's not like we can create constructor for v1, for v2 etc., as they'd be identical
might need some rework